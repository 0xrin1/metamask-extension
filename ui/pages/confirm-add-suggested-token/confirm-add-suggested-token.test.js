import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { fireEvent, screen } from '@testing-library/react';
import { acceptWatchAsset, rejectWatchAsset } from '../../store/actions';
import { renderWithProvider } from '../../../test/jest/rendering';
import mockState from '../../../test/data/mock-state.json';
import ConfirmAddSuggestedToken from '.';

const MOCK_SUGGESTED_ASSETS = [
  {
    id: 1,
    asset: {
      address: '0x8b175474e89094c44da98b954eedeac495271d0a',
      symbol: 'NEW',
      decimals: 18,
      image: 'metamask.svg',
      unlisted: false,
    },
  },
  {
    id: 2,
    asset: {
      address: '0xC8c77482e45F1F44dE1745F52C74426C631bDD51',
      symbol: '0XYX',
      decimals: 18,
      image: '0x.svg',
      unlisted: false,
    },
  },
];

jest.mock('../../store/actions', () => ({
  acceptWatchAsset: jest.fn().mockReturnValue({ type: 'test' }),
  rejectWatchAsset: jest.fn().mockReturnValue({ type: 'test' }),
}));

const renderComponent = (suggestedAssets = [...MOCK_SUGGESTED_ASSETS]) => {
  const baseStore = {
    metamask: {
      ...mockState.metamask,
      suggestedAssets,
    },
    history: {
      mostRecentOverviewPage: '/',
    },
  };
  const store = configureMockStore([thunk])(baseStore);
  return renderWithProvider(<ConfirmAddSuggestedToken />, store);
};

describe('ConfirmAddSuggestedToken Component', () => {
  it('should render', () => {
    renderComponent();

    expect(screen.getByText('Add Suggested Tokens')).toBeInTheDocument();
    expect(
      screen.getByText('Would you like to import these tokens?'),
    ).toBeInTheDocument();
    expect(screen.getByText('Token')).toBeInTheDocument();
    expect(screen.getByText('Balance')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'Cancel' })).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: 'Add Token' }),
    ).toBeInTheDocument();
  });

  it('should render the list of suggested tokens', () => {
    renderComponent();

    for (const { asset } of MOCK_SUGGESTED_ASSETS) {
      expect(screen.getByText(asset.symbol)).toBeInTheDocument();
    }
    expect(screen.getAllByRole('img')).toHaveLength(
      MOCK_SUGGESTED_ASSETS.length,
    );
  });

  it('should dispatch acceptWatchAsset when clicking the "Add Token" button', () => {
    renderComponent();
    const addTokenBtn = screen.getByRole('button', { name: 'Add Token' });

    fireEvent.click(addTokenBtn);
    expect(acceptWatchAsset).toHaveBeenCalled();
  });

  it('should dispatch rejectWatchAsset when clicking the "Cancel" button', () => {
    renderComponent();
    const cancelBtn = screen.getByRole('button', { name: 'Cancel' });

    expect(rejectWatchAsset).toHaveBeenCalledTimes(0);
    fireEvent.click(cancelBtn);
    expect(rejectWatchAsset).toHaveBeenCalledTimes(
      MOCK_SUGGESTED_ASSETS.length,
    );
  });

  describe('when the suggested token address matches an existing token address', () => {
    it('should show "already listed" warning', () => {
      const mockAssets = [
        {
          asset: {
            ...MOCK_SUGGESTED_ASSETS[0].asset,
            address: mockState.metamask.tokens[0].address,
          },
        },
      ];
      renderComponent(mockAssets);

      expect(
        screen.getByText(
          'This action will edit tokens that are already listed in your wallet, which can be used' +
            ' to phish you. Only approve if you are certain that you mean to change what these' +
            ' tokens represent. Learn more about',
        ),
      ).toBeInTheDocument();
      expect(
        screen.getByRole('link', { name: 'scams and security risks.' }),
      ).toBeInTheDocument();
    });
  });

  describe('when the suggested token symbol matches an existing token symbol and has a different address', () => {
    it('should show "reuses a symbol" warning', () => {
      const mockAssets = [
        {
          asset: {
            ...MOCK_SUGGESTED_ASSETS[0].asset,
            symbol: mockState.metamask.tokens[0].symbol,
          },
        },
      ];
      renderComponent(mockAssets);

      expect(
        screen.getByText(
          'A token here reuses a symbol from another token you watch, this can be confusing or deceptive.',
        ),
      ).toBeInTheDocument();
    });
  });
});
