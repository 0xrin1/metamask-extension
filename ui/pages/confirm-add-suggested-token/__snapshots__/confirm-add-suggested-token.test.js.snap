// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmAddSuggestedToken Component when clicking cancel button 1`] = `
<div>
  <div
    class="page-container"
  >
    <div
      class="page-container__header"
    >
      <div
        class="page-container__title"
      >
        Add Suggested Tokens
      </div>
      <div
        class="page-container__subtitle"
      >
        Would you like to import these tokens?
      </div>
      <div
        class="actionable-message actionable-message--warning actionable-message--with-right-button actionable-message--with-icon"
      >
        <svg
          viewBox="0 0 10 10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 0C2.2 0 0 2.2 0 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0 2c.4 0 .7.3.7.7s-.3.7-.7.7-.7-.2-.7-.6.3-.8.7-.8zm.7 6H4.3V4.3h1.5V8z"
            fill="#f8c000"
          />
        </svg>
        
        <div
          class="actionable-message__message"
        >
          A token here reuses a symbol from another token you watch, this can be confusing or deceptive.
        </div>
      </div>
    </div>
    <div
      class="page-container__content"
    >
      <div
        class="confirm-import-token"
      >
        <div
          class="confirm-import-token__header"
        >
          <div
            class="confirm-import-token__token"
          >
            Token
          </div>
          <div
            class="confirm-import-token__balance"
          >
            Balance
          </div>
        </div>
        <div
          class="confirm-import-token__token-list"
        >
          <div
            class="confirm-import-token__token-list-item"
          >
            <div
              class="confirm-import-token__token confirm-import-token__data"
            >
              <img
                alt=""
                class="identicon confirm-import-token__token-icon"
                src="metamark.svg"
                style="height: 48px; width: 48px; border-radius: 24px;"
              />
              <div
                class="confirm-import-token__name"
              >
                META
              </div>
            </div>
            <div
              class="confirm-import-token__balance"
            >
              <div
                class="currency-display-component"
                title="null"
              >
                <span
                  class="currency-display-component__text"
                />
              </div>
            </div>
          </div>
          <div
            class="confirm-import-token__token-list-item"
          >
            <div
              class="confirm-import-token__token confirm-import-token__data"
            >
              <img
                alt=""
                class="identicon confirm-import-token__token-icon"
                src="0x.svg"
                style="height: 48px; width: 48px; border-radius: 24px;"
              />
              <div
                class="confirm-import-token__name"
              >
                0X
              </div>
            </div>
            <div
              class="confirm-import-token__balance"
            >
              <div
                class="currency-display-component"
                title="null"
              >
                <span
                  class="currency-display-component__text"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="page-container__footer"
    >
      <footer>
        <button
          class="button btn--rounded btn-secondary btn--large page-container__footer-button"
          role="button"
          tabindex="0"
        >
          Cancel
        </button>
        <button
          class="button btn--rounded btn-primary btn--large page-container__footer-button"
          role="button"
          tabindex="0"
        >
          Add Token
        </button>
      </footer>
    </div>
  </div>
</div>
`;
